posit-dev/positron
# Handle supervisor exits more gracefully

This change causes Positron to handle an abrupt or unexpected exit of the supervisor process more gracefully. Before the change, the IDE basically became unresponsive. After the change, the situation is still Not Great, but it's recoverable; it's about the same as having all your interpreters crash at once. You'll see some error popups, but you can get back to a working state.

Because Positron has open websockets for all the active sessions, the best clue that the supervisor process has been terminated is that these sockets mysteriously all disconnect. So the main approach here is to fire disconnect events from the sessions; if we get one of these for a session that hasn't exited, we check the process table to see what's going on.

There are a couple of significant caveats:

- This approach only works when the supervisor process is local. We will probably want to implement a mechanism that works for remote supervisors too, if we add support for those.
- This approach won't help if the supervisor process is alive but unresponsive. Though that state hasn't occurred in nature so far, we will eventually want to guard against it.

Addresses https://github.com/posit-dev/positron/issues/5037.

### QA Notes

We don't expect supervisor exits to happen to anyone during the normal course of events, and this change does not attempt to make a supervisor exit a pleasant experience. 

This change also has a drive-by fix that addresses an issue running notebook kernels under the supervisor.
